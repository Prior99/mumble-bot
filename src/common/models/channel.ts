import { is, DataType, specify } from "hyrest";

import { MumbleUser } from ".";

export class Channel {
    @is()
    public name?: string;

    @is(DataType.int)
    public position?: number;

    @is() @specify(() => MumbleUser)
    public users?: MumbleUser[];

    @is() @specify(() => Channel)
    public children?: Channel[];
}
