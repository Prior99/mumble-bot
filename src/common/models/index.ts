export { Recording } from "./recording";
export { User } from "./user";
export { Label } from "./label";
export { CachedAudio } from "./cached-audio";
export { RecordingLabelRelation } from "./recording-label-relation";
export { DatabaseSound } from "./database-sound";
export { Dialog } from "./dialog";
export { DialogPart } from "./dialog-part";
export { MumbleLink } from "./mumble-link";
export { Permission } from "./permission";
export { PermissionAssociation } from "./permission-association";
export { Token } from "./token";
export { MumbleUser } from "./mumble-user";
export { Channel } from "./channel";
export { PlaybackOptions } from "./playback-options";
export { ForkOptions } from "./fork-options";
export { Setting } from "./setting";
