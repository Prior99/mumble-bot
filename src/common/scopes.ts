import { createScope } from "hyrest";

export const world = createScope();
export const login = createScope();
export const updateSound = createScope();
export const createLabel = createScope();
export const createPlaylist = createScope();
export const signup = createScope().include(login);
export const owner = createScope().include(world);
export const createSound = createScope();
export const createMumbleLink = createScope();
