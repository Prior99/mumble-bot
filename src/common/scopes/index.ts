import { createScope } from "hyrest";

export const world = createScope();
export const owner = createScope().include(world);
