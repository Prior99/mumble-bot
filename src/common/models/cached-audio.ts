import { is, scope, DataType, specify, uuid } from "hyrest";

import { world, enqueue, live, createSound } from "../scopes";

import { User } from ".";

/**
 * A cached audio.
 */
export class CachedAudio {
    constructor(id?: string, user?: User, duration?: number, date?: Date) {
        if (id && user && duration) {
            this.user = user;
            this.duration = duration;
            this.id = id;
        }
        if (date) {
            this.date = date;
        } else {
            this.date = new Date();
        }
    }

    /**
     * The date the audio was recorded.
     */
    @is() @scope(world, live) @specify(() => Date)
    public date?: Date;

    /**
     * The user from which the audio was recorded.
     */
    @is() @scope(world, live)
    public user?: User;

    /**
     * The id of the cached audio.
     */
    @scope(world, enqueue, live, createSound) @is().validate(uuid)
    public id?: string;

    /**
     * The duration of the audio in seconds.
     */
    @is(DataType.float) @scope(world, live)
    public duration?: number;
}
