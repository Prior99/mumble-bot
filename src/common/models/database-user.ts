import { Column, PrimaryGeneratedColumn, Entity, OneToMany } from "typeorm";
import { is, scope, specify, length, uuid, transform, only, required } from "hyrest";

import { world, login, owner, signup } from "../scopes";
import { hash } from "../utils";

import { Recording, PermissionAssociation, Token, Setting, MumbleLink } from ".";

/**
 * A user from the database.
 */
@Entity()
export class DatabaseUser {
    @PrimaryGeneratedColumn("uuid")
    @scope(world) @is().validate(uuid)
    public id?: string;

    /**
     * The username of this user.
     */
    @Column("varchar", { length: 100 })
    @is()
        .validate(length(3, 100), only(signup, required))
        .validateCtx(ctx => only(signup, value => ctx.validation.nameAvailable(value)))
    @scope(world, signup)
    public name?: string;

    /**
     * The password of the user.
     */
    @Column("varchar", { length: 200 })
    @transform(hash)
    @is().validate(length(8, 255)) @scope(login)
    public password?: string;

    @Column("varchar", { length: 200 })
    @is()
        .validate(length(8, 200), required)
        .validateCtx(ctx => only(signup, value => ctx.validation.emailAvailable(value)))
    @scope(login)
    public email?: string;

    @OneToMany(() => Recording, recording => recording.user)
    @is() @scope(world) @specify(() => Recording)
    public recordings?: Recording[];

    @OneToMany(() => Recording, recording => recording.reporter)
    @is() @scope(world) @specify(() => Recording)
    public reported?: Recording[];

    @OneToMany(() => PermissionAssociation, permissionAssociation => permissionAssociation.user)
    @is() @scope(world) @specify(() => PermissionAssociation)
    public permissionAssociations?: PermissionAssociation[];

    @OneToMany(() => Token, token => token.user)
    @is() @specify(() => Token)
    @scope(owner)
    public tokens?: Token[];

    /**
     * The custom settings of the user are stored key-value-wise in this object.
     */
    @OneToMany(() => Setting, setting => setting.user)
    @is() @specify(() => Setting)
    @scope(owner)
    public settings?: Setting[];

    @OneToMany(() => MumbleLink, mumbleLink => mumbleLink.user)
    @is() @specify(() => MumbleLink)
    public mumbleLinks?: MumbleLink;
}
