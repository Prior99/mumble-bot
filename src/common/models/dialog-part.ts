import { PrimaryGeneratedColumn, Entity, OneToMany } from "typeorm";
import { is, scope, uuid } from "hyrest";

import { world, createDialog } from "../scopes";

import { Dialog, Recording  } from ".";

@Entity()
export class DialogPart {
    @PrimaryGeneratedColumn("uuid")
    @scope(world) @is().validate(uuid)
    public id?: string;

    @OneToMany(() => Recording, recording => recording.dialogParts)
    @is() @scope(world, createDialog)
    public recording?: Recording;

    @OneToMany(() => Dialog, dialog => dialog.parts)
    @is() @scope(world)
    public dialog?: Dialog;
}
