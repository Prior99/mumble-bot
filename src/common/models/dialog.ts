import { Column, PrimaryGeneratedColumn, Entity, ManyToOne, CreateDateColumn } from "typeorm";
import { is, scope, specify, uuid, DataType } from "hyrest";
import { world, createDialog } from "../scopes";
import { DialogPart } from ".";

/**
 * A dialog as represented in the database including all its records.
 * @typedef Dialog
 */
@Entity()
export class Dialog {
    /**
     * Unique id of this dialog.
     */
    @PrimaryGeneratedColumn("uuid")
    @scope(world) @is().validate(uuid)
    public id?: number;

    /**
     * The date when this dialog was submitted.
     */
    @CreateDateColumn()
    @scope(world) @is() @specify(() => Date)
    public submitted?: Date;

    /**
     * How often this dialog was used.
     */
    @Column("int", { default: 0 })
    @is(DataType.int) @scope(world)
    public used?: number;

    /**
     * All records belonging to this dialog.
     */
    @ManyToOne(() => DialogPart, dialogPart => dialogPart.dialog)
    @is() @scope(world, createDialog) @specify(() => DialogPart)
    public parts?: DialogPart[];
}
