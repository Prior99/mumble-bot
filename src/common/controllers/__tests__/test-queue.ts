import { copy } from "fs-extra";
import { fft, util } from "fft-js";
import { populate } from "hyrest";
import { Connection } from "mumble";
import {
    api,
    createPlaylist,
    createUserWithToken,
    createSoundWithCreatorAndSpeaker,
    startDb,
    stopDb,
} from "../../../test-utils";
import { world } from "../../scopes";
import { Token, User, Sound, CachedAudio } from "../../models";
import { AudioOutput, AudioCache } from "../../../server";
import { ServerConfig } from "../../../config";

describe("queue controller", () => {
    let token: Token;
    let user: User;
    let sound: Sound;
    let mumble: Connection;
    let audioOutput: AudioOutput;
    let voice: number[];

    beforeEach(startDb);
    afterEach(stopDb);

    beforeEach(async () => {
        const userAndToken = await createUserWithToken();
        token = userAndToken.token;
        user = userAndToken.user;
        sound = (await createSoundWithCreatorAndSpeaker({ duration: 1 } as Sound)).sound;
        await copy(
            `${__dirname}/../../../__fixtures__/sin-short.mp3`,
            `${tsdi.get(ServerConfig).soundsDir}/${sound.id}`,
        );
        mumble = tsdi.get("MumbleConnection");
        audioOutput = tsdi.get(AudioOutput);

        voice = [];
        mumble.inputStream().on("data", (data: Buffer) => {
            for (let i = 0; i < data.length; i += 2) {
                const num = data.readInt16LE(i);
                voice.push((num / Math.pow(2, 16)) * 2 - 1);
            }
        });
    });

    describe("POST /queue", () => {
        it("responds 401 without a valid token", async () => {
            const response = await api().post("/queue");
            expect(response.body).toEqual({ message: "Unauthorized." });
            expect(response.status).toBe(401);
        });

        describe(`with type="sound"`, () => {
            it("enqueues a sound", async done => {
                const requestQueueItem = {
                    type: "sound",
                    sound: { id: sound.id },
                    pitch: 50,
                };
                const response = await api().post(`/queue`)
                    .set("authorization", `Bearer ${token.id}`)
                    .send(requestQueueItem);
                expect(response.status).toBe(201);
                const resultQueueItem = {
                    ...requestQueueItem,
                    user: { id: user.id },
                };
                expect(response.body).toMatchObject({ data: resultQueueItem });
                const updatedSoundResponse = await api().get(`/sound/${sound.id}`)
                    .set("authorization", `Bearer ${token.id}`);
                expect(updatedSoundResponse.body.data.used).toBe(sound.used + 1);
                audioOutput.once("shift", item => {
                    expect(voice.length).toMatchSnapshot();
                    expect(item).toMatchObject(resultQueueItem);
                    done();
                });
            });
        });

        describe(`with type="cached audio"`, () => {
            it("enqueues a cached audio", async done => {
                const cachedAudio = {
                    date: new Date("2018-11-15Z10:00:00"),
                    user,
                    id: "0edf2ba3-d372-4ccf-8b85-f475423747cb",
                    duration: 15,
                    amplitude: 38,
                };
                await copy(
                    `${__dirname}/../../../__fixtures__/sin-short.mp3`,
                    `${tsdi.get(ServerConfig).tmpDir}/${cachedAudio.id}`,
                );
                await tsdi.get(AudioCache).add(populate(world, CachedAudio, cachedAudio));
                const requestQueueItem = {
                    type: "cached audio",
                    cachedAudio: { id: cachedAudio.id },
                    pitch: 50,
                };
                const response = await api().post(`/queue`)
                    .set("authorization", `Bearer ${token.id}`)
                    .send(requestQueueItem);
                expect(response.status).toBe(201);
                const resultQueueItem = {
                    ...requestQueueItem,
                    user: { id: user.id },
                };
                expect(response.body).toMatchObject({ data: resultQueueItem });
                audioOutput.once("shift", item => {
                    expect(voice.length).toMatchSnapshot();
                    expect(item).toMatchObject(resultQueueItem);
                    done();
                });
            });
        });

        describe(`with type="playlist"`, () => {
            it("enqueues a playlist", async done => {
                const playlist = await createPlaylist(user, 0, sound, sound);
                const requestQueueItem = {
                    type: "playlist",
                    playlist: { id: playlist.id },
                    pitch: 50,
                };
                const response = await api().post(`/queue`)
                    .set("authorization", `Bearer ${token.id}`)
                    .send(requestQueueItem);
                expect(response.status).toBe(201);
                const resultQueueItem = {
                    ...requestQueueItem,
                    user: { id: user.id },
                };
                expect(response.body).toMatchObject({ data: resultQueueItem });
                audioOutput.once("shift", item => {
                    expect(voice.length).toMatchSnapshot();
                    expect(item).toMatchObject(resultQueueItem);
                    done();
                });
            });

            [-200, 600].forEach(pitch => {
                it(`enqueues a playlist with pitch ${pitch}`, async done => {
                    const playlist = await createPlaylist(user, pitch, sound, sound);
                    const requestQueueItem = {
                        type: "playlist",
                        playlist: { id: playlist.id },
                    };
                    const response = await api().post(`/queue`)
                        .set("authorization", `Bearer ${token.id}`)
                        .send(requestQueueItem);
                    expect(response.status).toBe(201);
                    const resultQueueItem = {
                        ...requestQueueItem,
                        user: { id: user.id },
                    };
                    expect(response.body).toMatchObject({ data: resultQueueItem });
                    audioOutput.once("shift", item => {
                        expect(voice.length).toMatchSnapshot();
                        const phasors = fft(voice.slice(0, Math.pow(2, 15)));
                        const frequencies = util.fftFreq(phasors, 48000);
                        const magnitudes = util.fftMag(phasors);
                        let highest: number;
                        let highestIndex: number;
                        for (let i = 1; i < magnitudes.length; ++i) {
                            if (highest === undefined || magnitudes[i] > highest) {
                                highest = magnitudes[i];
                                highestIndex = i;
                            }
                        }
                        expect({
                            magnitude: Math.round(highest),
                            frequency: Math.round(frequencies[highestIndex]),
                        }).toMatchSnapshot();
                        expect(item).toMatchObject(resultQueueItem);
                        done();
                    });
                });
            });
        });
    });
});
