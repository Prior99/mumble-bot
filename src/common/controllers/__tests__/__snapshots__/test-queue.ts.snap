// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queue controller POST /queue with type="cached audio" enqueues a cached audio 1`] = `48033`;

exports[`queue controller POST /queue with type="playlist" enqueues a playlist 1`] = `96066`;

exports[`queue controller POST /queue with type="playlist" enqueues a playlist with pitch -200 1`] = `96066`;

exports[`queue controller POST /queue with type="playlist" enqueues a playlist with pitch -200 2`] = `
Object {
  "frequency": 201,
  "magnitude": 5657,
}
`;

exports[`queue controller POST /queue with type="playlist" enqueues a playlist with pitch 600 1`] = `96066`;

exports[`queue controller POST /queue with type="playlist" enqueues a playlist with pitch 600 2`] = `
Object {
  "frequency": 319,
  "magnitude": 7658,
}
`;

exports[`queue controller POST /queue with type="sound" enqueues a sound 1`] = `48033`;
